### PR TITLE
Vendor dependencies

### LINKED ISSUE #1097: Checking in vendored dependencies?
(I couldn't find any previous issues / pull requests on the following request, so I'm opening this one.)

I was wondering if the Traefik maintainers were open to checking in vendored dependencies into git. Right now, in order to build the project, developers first have to do `glide install` to bring in dependencies in the correct versions. For me, this is multiple unfavorable implications:

- Users need to have glide installed just to build the project (and ideally use the right version).
- Build times are significantly prolonged. While this is not so much of a problem when building locally (as dependencies shouldn't change too often, but see my other points), I noticed that large portions of CI time are consumed by installing dependencies. We seem to be speaking minute scale here.
- Developers need to run `glide install` on every git pull to guarantee consistency. One problem with this is that it's easy to forget and could lead to hard-to-trace bugs. The other concern is that you need to do it on every branch switch just to be safe.
- The build can break because of dependencies gone away, being renamed, Github failing, etc. Coincidentally, on my first Traefik PR, a ran into a Github glitch that caused my Travis build to fail.

All of these problems would go away if the `vendor/` folder was checked into the VCS.

Of course, there are also downsides:

- Repo size increases. While this is a problem when using glide as-is, the popular glide plugin [glide-vs](https://github.com/sgotti/glide-vc) can reduce the size significantly: With the current master, the  vendor directory can be trimmed down from more than 300 MB to just 60 MB. Another 1-2 MB can be shaved off by also removing licence files, but that's [somewhat controversial](http://engineeredweb.com/blog/2016/go-why-not-strip-unused-pkgs/).
- PRs with dependency changes get harder to read since third-party and application sources are hard to distinguish in Github's PR UI. One way to deal with this is to isolate dependency changes into separate commits and review commit-wise. Only few PRs should be affected by this since dependency-changing PRs are usually in the minority.

I realize that there are numerous proponents for each of the commit/don't commit camps, and obviously I'm part of the commit camp. For me, the benefits clearly outnumber of drawbacks. I'd be curious to know how the Traefik project thinks about this matter and whether there's interest in "switching camps". :-)

Thanks!

### PR DESCRIPTION
**Pull request by timoreimann**
This PR switches the way we handle dependencies by committing the vendor folders (main and integration) into the VCS.

Some noteworthy changes:

- Committing the vendor folders populated by `glide install [--strip-vendor]` as-is results in more than 400 MB of dependency files. That's why we also `--skip-test`s and run [glide-vc](https://github.com/sgotti/glide-vc) afterwards to trim down deps to 62 MB. (Note that I have decided to keep the licencing information in place for reasons [described here](http://engineeredweb.com/blog/2016/go-why-not-strip-unused-pkgs/).)
- We can only strip vendor folders with the main dependencies; the test build breaks if we try to do the same on the integration deps.
- The latest versions of glide-vc use `glide list` to determine the files which can be eliminated. What I found out though is that this even removes needed files. This might be a bug in glide; for now, I have resorted to using the glide.lock file as the basis for what can be trimmed (`--use-glide-lock` switch on glide-vc).
- All of this seems too complicated to be done correctly by a human. Hence, I created the helper/wrapper script `glide.sh` that should be used whenever we need to work with dependencies (install/update/get/trim).

I separated the vendor folder changes into dedicated commits so that the actual changes to the project infrastructure can be reviewed more easily (the first commit).

Fixes #1097.

### CODE REVIEW

**Comment by timoreimann** [comment]
NB: I'm currently having issues making the Travis CI run complete in time: The integration tests keep failing and must be retried until they pass. Once that happens, there's not enough time left to finish cross-compiling within the 50 minute timeout that Travis imposes. I'm not entirely sure whether this is a consequence of my PR or just flakiness, but I'm slightly leaning towards the latter. I'd be glad for some feedback on this part too.

**Comment by timoreimann** [comment]
Looks like it went green on the first try. :-)

@emilevauge @errm @vdemeester: PTAL.

**Comment by emilevauge** [comment]
Wow! Great job @timoreimann !
This is not a full review for now.
I have a major concern on this:
>Committing the vendor folders populated by glide install [--strip-vendor] as-is results in more than 500 MB of dependency files. That's why we also --skip-tests and run glide-vc afterwards to trim down deps to 62 MB. (Note that I have decided to keep the licencing information in place for reasons described here.)

I don't think this respect deps' licences sadly. I know this is annoying, and I'm not a lawyer, but I think we can't modify what in `vendor/`...

**Comment by timoreimann** [comment]
@emilevauge 

Disclaimer: I'm not a lawyer either. :-)

Yes, that's indeed somewhat of a disputed topic. FWIW, the [Google lawyers think it's okay](https://github.com/Masterminds/glide/issues/303#issuecomment-197598334), and govendor follows a similar approach by just holding on to *.go and licencing files by default. It'll be interesting to see what conclusions golang/dep#120 will eventually draw in this regard.

That said, I realize that neither of the mentioned entities will defend you, me, or the Traefik project when in doubt. :-) Therefore, I'll push another commit to remove the contentious part.

**Comment by emilevauge** [comment]
On this, WDYT @containous/traefik ?

**Comment by timoreimann** [comment]
Travis CI fails right now due to the new `validate-vendor` claiming there's a file I didn't check in. However, I see it being committed locally.

Will need to investigate.

**Comment by timoreimann** [comment]
For some reason, the build container behaves slightly differently than my host system: Running `glide install --strip-vendor` during the `validate-vendor` script inside the container produces one vendored file `vendor/bitbucket.org/ww/goautoneg/.hg_archival.txt` diffing by one line:

```
root@b483094ac340:/go/src/github.com/containous/traefik# git diff
diff --git a/vendor/bitbucket.org/ww/goautoneg/.hg_archival.txt b/vendor/bitbucket.org/ww/goautoneg/.hg_archival.txt
index b9a2ff9..3c3827a 100644
--- a/vendor/bitbucket.org/ww/goautoneg/.hg_archival.txt
+++ b/vendor/bitbucket.org/ww/goautoneg/.hg_archival.txt
@@ -3,4 +3,3 @@ node: 75cd24fc2f2c2a2088577d12123ddee5f54e0675
 branch: default
 latesttag: null
 latesttagdistance: 5
-changessincelatesttag: 5
```

If I install the dependencies on my host machine (Mac) using glide, that last line is there in the file [as can be seen on the committed file](https://github.com/timoreimann/traefik/blob/65676b099c862a9f1c35f8ab6685551efc74658d/vendor/bitbucket.org/ww/goautoneg/.hg_archival.txt).

I suspected different glide versions but they are equal (except for the varying architectures).

I have been debugging this issue for a few hours now and start to run out of things to try next. If anyone has an idea, I'd be happy to hear.

**Comment by errm** [comment]
sounds like something to do with the version of hg(mercurial)?

Is there a difference in the output of `hg version`?

**Comment by timoreimann** [comment]
@errm good point: I have 3.9 on my host, the jessie-backed golang Docker image is on a fairly old 3.1.

There's 3.9 available on jessie-backports. I'll try to upgrade to that one in the container.

Thanks!

**Comment by timoreimann** [comment]
Looks like it was the mercurial version. :tada:

Validation completes successfully now. Crossing fingers for the Travis CI run now...

**Comment by timoreimann** [comment]
Build went green. 😀

I think we are ready for review again.

(if / once this PR gets approval, please do not merge yourself; I'd like to squash commits first.)

**Comment by timoreimann** [comment]
@emilevauge @vdemeester @errm any chance to get a review and ideally an LGTM? :-)

**Comment by klausenbusk** [comment]
From a outsider just looking around :). Is there any reason you need to add all the files, couldn't you just use something like submodule?

**Comment by vdemeester** [comment]
@klausenbusk using git submodule is the same as not vendoring, and has shortcommings too 

- as for not vendoring, if a repository is gone, you can't build anymore
- if the repository is *not* a git repository, you are screwed 👼 

The thing with vendoring is : do we clean the mess or not — i.e. do we only keep packages we use, without the test files, and keeping LICENCE and other important files (and limit a lot what we store) ; or do we not clean and store waaayy too much file. 

I'm for the first (and there is a few project who do that).

**Comment by klausenbusk** [comment]
> @klausenbusk using git submodule is the same as not vendoring, and has shortcommings too

Oh yeh, you correct. I didn't through of that, for my small "hobby"-project (a few hundred lines) I just use submodules, but is has some shortcoming defiantly.

**Comment by timoreimann** [comment]
@vdemeester since you're around anyway :) Any chance for a review of my PR? Thanks!

**Comment by vdemeester** [comment]
I'm in 💯 on vendoring, but I'm really not a huge fan of `glide` and I really think we should clean the deps (strip vendor, remove test files and package that are not used — not sure how glide manage all of these). I understand removing some files from `vendor` might no respect some license and that's why I don't like transitive dependencies either (because you don't *control* those).

So, from me it would be a LGTM with removing some file to keep only what's needed. But I'm a little biased on the subject (@containous/traefik maintainers knows my point of view 😅) so I'll let other maintainer give their opinion.

**Comment by timoreimann** [comment]
I think I can answer those glide behavior questions:

- vendor stripping: Happens when you pass the `-v` parameter. Traefik already does this for the main repo but not for the integration one. (But even if we wanted to, it doesn't work for the latter right now. See my PR description comment.)
- test file removal: Needs an extra switch. My initial version of this PR had it and it worked fine.
- Removing unused files: Needs the glide-vc (vendor cleaning) plugin, glide only removes whole unused package dependencies. Again, the initial PR commits contain it.

**Comment by timoreimann** [comment]
Following my discussion with @emile, this PR is back to where it used to be in the beginning, with dependency pruning enabled. The overall `{integration}/vendor` folder size is 62 MB.

Do we agree to move forward with vendoring as-is? @vdemeester?

File: build.Dockerfile [hunk 1]
```diff
@@ -1,5 +1,12 @@
 FROM golang:1.7
 
+# Install a more recent version of mercurial to avoid mismatching results
+# between glide run on a decently updated host system and the build container.
```
Comment (emilevauge) [inline build.Dockerfile]: Oh god :cry:

File: .gitignore [hunk 2]
```diff
@@ -1,15 +1,13 @@
 /dist
-gen.go
+/autogen/gen.go
```
Comment (emilevauge) [inline .gitignore]: I think we could now include generated files in git. This will allow to `go get` Traefik. WDYT? But that wan be done in another PR.

Comment (timoreimann) [inline .gitignore]: I'd love to get rid of the extra generator step. I haven't worked a lot with generated code, however, so no idea if checking that in is a good practice or not.

(I know the Kubernetes folks are trying hard to undo committing generated code, but that might be because they are playing in a different league.)

Either way, discussing / considering in a separate issue / PR sounds good. 👍

**Comment by timoreimann** [comment]
@vdemeester with pruning now part of this PR, can I get your LGTM? :)

File: build.Dockerfile (hunk 1 above)
Comment (timoreimann) [inline build.Dockerfile]: The golang builder is based on Debian jessie. 🙁

**Comment by emilevauge** [comment]
Ping @containous/traefik
